### PR TITLE
Add region data to JSON report (fix #1793)

### LINKED ISSUE #1793: A way to get report data in detail
The report method only returns the total coverage amount. However, sometimes we need this information in detail. For example, I need to know the function-by-function coverage of a file, and I can't do this using the report method. I know it's possible to do this using the html report method, but having the value inside a variable would make it much easier for a more detailed analysis of where the coverage is bad.

So if the report method returned a dictionary, for example, with the names of the functions and their coverage, it would be very useful.

### PR DESCRIPTION
**Pull request by devdanzin**
This PR addresses #1793 by adding data regarding class and function/method coverage to the JSON report.

### CODE REVIEW

File: coverage/jsonreport.py [hunk 1]
```diff
@@ -122,6 +126,45 @@ def report_one_file(self, coverage_data: CoverageData, analysis: Analysis) -> di
...
+            }
+            if self.config.json_show_contexts:
+                contexts = coverage_data.contexts_by_lineno(narrowed_analysis.filename)
+                reported_file[region.kind][region.name]["contexts"] = contexts
+            if coverage_data.has_arcs():
+                narrowed_summary.update({
+                    "num_branches": narrowed_nums.n_branches,
+                    "num_partial_branches": narrowed_nums.n_partial_branches,
+                    "covered_branches": narrowed_nums.n_executed_branches,
+                    "missing_branches": narrowed_nums.n_missing_branches,
```
Comment (nedbat) [inline coverage/jsonreport.py:160]: When I change this to `"xmissing_branches"`, no tests fail, so it seems like we're missing a test of regions with branches?

Comment (nedbat) [reply coverage/jsonreport.py:160]: I can take over finishing this.

Comment (devdanzin) [reply coverage/jsonreport.py:160]: Feel free to, I probably won't have time to work on it today and would love to see it land.

**Comment by nedbat** [comment]
This work is finished in commit ab609ef0fb235454050cf8383427ce5f1b0ec8e9.

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.6.0](https://pypi.org/project/coverage/7.6.0).

**Pull request closed without merging**